src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c

```diff
@@ -792,8 +792,9 @@ zocl_xclbin_ctx(struct drm_zocl_dev *zdev, struct drm_zocl_ctx *ctx,
 
 		if (cu_idx != ZOCL_CTX_VIRT_CU_INDEX) {
 			/* Try clear exclusive CU */
-			ret = test_and_clear_bit(cu_idx, client->excus);
-			if (!ret)
+			ret = test_and_clear_bit(cu_idx, client->excus) ?
+                                  0 : -EINVAL;
+			if (ret)
```

Comment:
It is a little bit weird of using "-EINVAL" here. The "-EINVAL" does not really play a role. It could be any non-zero value. What you really want to to do is to reverse the return value of test_and_clear_bit so that we don't hit the second if statement. I would suggest a different approach, like
if (cu_idx != ZOCL_CTX_VIRT_CU_INDEX) {
        /* Try clear exclusive CU */
        ret = test_and_clear_bit(cu_idx, client->excus);
        if (!ret) {
                /* Maybe it is shared CU */
                ret = test_and_clear_bit(cu_idx, client->shcus) ?
                                0 : -EINVAL;
                if (ret) {
                        DRM_ERROR("can not remove unreserved cu");
                        goto out;
                }
        }
}
or, I prefer the following fix which could scale if we have other CU types.
if (cu_idx != ZOCL_CTX_VIRT_CU_INDEX) {
        /* Try clear exclusive CU */
        ret = test_and_clear_bit(cu_idx, client->excus);
        if (!ret) {
                /* Maybe it is shared CU */
                ret = test_and_clear_bit(cu_idx, client->shcus);
        }
        if (!ret) {
                DRM_ERROR("can not remove unreserved cu");
                ret = -EINVAL;
                goto out;
        }
}

Please let me know what do you think. Thanks!

Reply:
In the 1st method it will again hit DRM_ERROR case if it is exclusive cu. Second method is perfect, I will update and push the change. Thanks larry.

Reply:
changed the code as suggested, please review